clang/lib/Sema/SemaStmt.cpp

```diff
@@ -971,6 +971,35 @@ StmtResult Sema::ActOnIfStmt(SourceLocation IfLoc,
   if (!ConstevalOrNegatedConsteval && !elseStmt)
     DiagnoseEmptyStmtBody(RParenLoc, thenStmt, diag::warn_empty_if_body);
 
+  // Checks for if condition variable usage in else scope
+  if (elseStmt) {
```

Comment:
Hmm.. This whole thing is a little too pattern matchy for me, and without a visitor is now going to just do this the 'hard' way.  

I think `ActOnIfStmt` is probably the wrong place for this entirely, its too late at this point.  We should probably do it somewhere in `ActOnVariableDeclarator` (or something like that?) and check the scope.  `Scope` itself doesn't do diagnostics, but will have the context that we are 'in' a for loop, and can store the condition variable.  

As this is written, an `if-else` chain ends up having the last `else` body checked N times.  In addition to everything else, thats just wrong.

Instead, when we see a var-reference being created, we should be checking all the active `if` statements, and be variables that we've stored along the way.